pkg/scheduler/algorithm/predicates/csi_volume_predicate.go

```diff
@@ -129,28 +134,62 @@ func (c *CSIMaxVolumeLimitChecker) filterAttachableVolumes(
 			continue
 		}
 
-		pvName := pvc.Spec.VolumeName
-		// TODO - the actual handling of unbound PVCs will be fixed by late binding design.
-		if pvName == "" {
-			klog.V(4).Infof("Persistent volume had no name for claim %s/%s", namespace, pvcName)
+		driverName, volumeHandle := c.getCSIDriver(pvc)
+		// if we can't find driver name or volume handle - we don't count this volume.
+		if driverName == "" || volumeHandle == "" {
 			continue
 		}
-		pv, err := c.pvInfo.GetPersistentVolumeInfo(pvName)
-
-		if err != nil {
-			klog.V(4).Infof("Unable to look up PV info for PVC %s/%s and PV %s", namespace, pvcName, pvName)
-			continue
-		}
-
-		csiSource := pv.Spec.PersistentVolumeSource.CSI
-		if csiSource == nil {
-			klog.V(4).Infof("Not considering non-CSI volume %s/%s", namespace, pvcName)
-			continue
-		}
-		driverName := csiSource.Driver
 		volumeLimitKey := volumeutil.GetCSIAttachLimitKey(driverName)
-		result[csiSource.VolumeHandle] = volumeLimitKey
+		result[volumeHandle] = volumeLimitKey
 
 	}
 	return nil
 }
+
+func (c *CSIMaxVolumeLimitChecker) getCSIDriver(pvc *v1.PersistentVolumeClaim) (string, string) {
+	pvName := pvc.Spec.VolumeName
+	namespace := pvc.Namespace
+	pvcName := pvc.Name
+
+	placeHolderCSIDriver := ""
+	placeHolderHandle := ""
+	if pvName == "" {
+		klog.V(4).Infof("Persistent volume had no name for claim %s/%s", namespace, pvcName)
+		return c.getDriverNameFromSC(pvc)
+	}
+	pv, err := c.pvInfo.GetPersistentVolumeInfo(pvName)
+
+	if err != nil {
+		klog.V(4).Infof("Unable to look up PV info for PVC %s/%s and PV %s", namespace, pvcName, pvName)
```

Comment:
Can you add a comment here that this is an error condition where either:
* User prebound the PVC to a non-existing PV
* PV object got deleted erroneously

Should we fallback to SC.provisionername?

Reply:
Don't we have a pv proection controller to prevent things like PV object getting deleted erroneously? First one is more likely perhaps - but then it seems like no pod can start using such a PVC.

Reply:
I'm trying to think of the case where someone bypasses the scheduler by directly setting Pod.nodename and they have prebound the PVC and the PV gets bound in the middle.  Then we won't count it.  But the same issue also exists if they have an empty or non-existent storageclass.  So I'm not sure if we'll be able to fully solve this race.

Reply:
So if they prebound the PVC then most likely it will have missing SC too right? Although likely-hood of binding happening in middle for prebound PVCs is small. I am for leaving this code as it is. I think this is "good middleground".

Reply:
Can we require that users create a StorageClass object with correct provisioner name even for prebinding PVC/PV, in order for the volume counting to work correctly?

Reply:
I am not sure. So if a PVC is not bound and does not have a storageclass with delayed binding (because user is trying to submit pv/pvc/pod template at once) - won't the scheduler not schedule the pod until PVC is bound?  That seems to be my experience and in which case - for pre-bound PVCs this will be much less of a problem. Am I missing something?

Reply:
added fallback.